### PR TITLE
[MLA] Tuned params mapping for Mistral-Small-4 on TPU v7x

### PR DESCRIPTION
**Pull request by karan**
# Description

This PR introduces optimized, hardware-tuned MLA parameters for Mistral-Small-4 on TPU v7x.

In addition to adding the tuned parameters, this PR fixes a test infrastructure issue where tests with identical default attributes generated duplicate test names when missing specific key dimensions, preventing comprehensive testing.

Previously, the `mla_tuned_vs_baseline_test.py` benchmark was executing the Pallas kernel eagerly in Python without a JIT compiler decorator. As a result, the benchmark was primarily measuring Python dispatch and object creation overhead rather than true hardware latency, leading to wildly inflated "speedups" (e.g., 10-30%) for tuned parameters.

This PR fixes the test by wrapping the `run_kernel` function with `@jax.jit(static_argnames=[...])`. This ensures that XLA properly compiles the execution block, allowing us to benchmark the true on-chip TPU latency.

This is similar to #2831.

### Context & Problem Solved

Previously, serving Mistral-Small-4 on TPU v7x resulted in JAX falling back to default parameter heuristics. This triggered warnings during server initialization and missed out on optimal hardware execution layouts.

```
(EngineCore pid=345212) WARNING 06-17 20:48:52 [tuned_params.py:383] No tuned parameters found for the given tuning key: TuningKey(case='batched_decode', max_num_tokens=4, actual_num_q_heads=32, actual_lkv_dim=256, actual_r_dim=64, kv_dtype='float8_e4m3fn', q_dtype='float8_e4m3fn', total_num_pages=1506, page_size_per_kv_packing=256, kv_packing=4, max_num_seqs=160, pages_per_seq=9, s_dtype='bfloat16', soft_cap=None, chunk_prefill_size=None, sliding_window=None, p_same_dtype_as_v=True), using default parameters
```

Tuning keys spanning across 32 and 128 actual query heads and varying token contexts have now been explicitly mapped for 1506 pages.

```
python -m tools.kernel.tuner.v1.kernel_tuner_runner --kernel_tuner_name=mla_kernel_tuner --run_locally=True --case_set_id=mla_local_run_mistral4_50iters --run_id=0 --case_set_desc='MLA local tuning run for Mistral-Small-4 with 50 iters' --tpu_version=tpu7x --tpu_cores=8 --mla_actual_num_q_heads=32 --mla_actual_lkv_dim=256
```

# Tests

```bash
python tools/kernel/tuner/v1/inspect_result_cli.py --source=local --db-path=/tmp/kernel_tuner_run_2026_06_23_21_05_14 query_min_latency --case_set_id=mla_local_run_mistral4_50iters --run_id=0
```

```bash
cd ~/tpu-inference
python3 tools/kernel/tuner/v1/inspect_result_cli.py --source=local --db-path=/tmp/kernel_tuner_run_2026_06_24_20_28_12 query_min_latency --case_set_id=mla_local_run_mistral4_1000iters --run_id=0
```

```
case_id  latency_us  warmup_us  case            max_num_tokens  actual_num_q_heads  actual_lkv_dim  actual_r_dim  kv_dtype       q_dtype        total_num_pages  page_size_per_kv_packing  kv_packing  max_num_seqs  pages_per_seq  s_dtype   soft_cap  chunk_prefill_size  sliding_window  p_same_dtype_as_v  decode_batch_size  num_kv_pages_per_block  num_queries_per_block  vmem_limit_bytes
-------  ----------  ---------  --------------  --------------  ------------------  --------------  ------------  -------------  -------------  ---------------  ------------------------  ----------  ------------  -------------  --------  --------  ------------------  --------------  -----------------  -----------------  ----------------------  ---------------------  ----------------
405      603         638        batched_decode  1024            32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               8                  3                       1                      62914560        
241      604         632        batched_decode  128             32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               8                  3                       1                      62914560        
119      597         627        batched_decode  16              32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               16                 3                       1                      62914560        
246      609         79994      batched_decode  160             32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               4                  3                       1                      67108864        
446      598         651        batched_decode  2048            32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               8                  3                       1                      62914560        
323      604         683        batched_decode  256             32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               8                  3                       1                      62914560        
123      607         80736      batched_decode  32              32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               4                  3                       1                      67108864        
36       601         4477862    batched_decode  4               32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               8                  3                       1                      62914560        
328      600         81430      batched_decode  512             32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               4                  3                       1                      67108864        
200      596         710        batched_decode  64              32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               8                  3                       1                      62914560        
41       604         86971      batched_decode  8               32                  256             64            float8_e4m3fn  float8_e4m3fn  1506             256                       4           160           9              bfloat16  None      None                None            True               4                  3                       1                      67108864        
```

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/kernels/mla/v2/tuned_params.py [hunk 1]
```diff
@@ -373,6 +373,149 @@ class TunableParams:
         num_queries_per_block=1,
         vmem_limit_bytes=62914560,
     ),
+    TuningKey(
+        case="batched_decode",
+        max_num_tokens=4,
+        actual_num_q_heads=32,
+        actual_lkv_dim=256,
+        actual_r_dim=64,
```
Comment (BirdsOfAFthr) [inline tpu_inference/kernels/mla/v2/tuned_params.py:440]: can you tell me why we do not need to set
```
        total_num_page,
        page_size_per_kv_packing,
        kv_packing,
        max_num_seqs,
        pages_per_seq,
```

Comment (karan) [reply tpu_inference/kernels/mla/v2/tuned_params.py:440]: good q.

Basically we're using [the defaults](https://github.com/vllm-project/tpu-inference/blob/913ece8683c7dedb84664edfa74e69b21b7f6759/tpu_inference/kernels/mla/v2/tuned_params.py#L31-L35). And they are aligned with what I saw during server bringup:

```
(EngineCore pid=345212) WARNING 06-17 20:48:52 [tuned_params.py:383] No tuned parameters found for the given tuning key: TuningKey(case='batched_decode', max_num_tokens=4, actual_num_q_heads=32, actual_lkv_dim=256, actual_r_dim=64, kv_dtype='float8_e4m3fn', q_dtype='float8_e4m3fn', total_num_pages=1506, page_size_per_kv_packing=256, kv_packing=4, max_num_seqs=160, pages_per_seq=9, s_dtype='bfloat16', soft_cap=None, chunk_prefill_size=None, sliding_window=None, p_same_dtype_as_v=True), using default parameters
```

How did you decide to override these for the large model?

Comment (BirdsOfAFthr) [reply tpu_inference/kernels/mla/v2/tuned_params.py:440]: so you are still seeing tuned params not found? Essentially you are still using defaults? That might be why you did not see any speedup.

It is expected if you did not specify the 5 params I mentioned, because there are defaults for those.

> How did you decide to override these for the large model?

tagged you in the doc.

Comment (karan) [reply tpu_inference/kernels/mla/v2/tuned_params.py:440]: > so you are still seeing tuned params not found? 

Nope - it's using these tuned values now.

Comment (BirdsOfAFthr) [reply tpu_inference/kernels/mla/v2/tuned_params.py:440]: These keys have [defaults](https://github.com/vllm-project/tpu-inference/blob/913ece8683c7dedb84664edfa74e69b21b7f6759/tpu_inference/kernels/mla/v2/tuned_params.py#L29-L35). Some of those values are different because those hard-coded numbers are for the DSv3 benchmarks configs. At least for small Mistral, the `max_num_seqs` is 16. `page_size_per_kv_packing` and `pages_per_seq` should be different too. So I am not sure how we can find the tuned params. Did you add the logging that I was referring to in the attn inference?

Comment (karan) [reply tpu_inference/kernels/mla/v2/tuned_params.py:440]: Thanks for pointing me in the right direction!

I added the explicit parameter shape logging natively in attention_interface.py as you recommended and saw:

> actual_num_q_heads=32, actual_lkv_dim=256, actual_r_dim=64, page_size_per_kv_packing=32, kv_packing=32, max_num_seqs=16, pages_per_seq=3

I then moved the optimal tile block distributions directly into
tuned_params.py for those exact parameter dimensions covering batched_decode.

File: tpu_inference/layers/common/attention_interface.py
Comment (BirdsOfAFthr) [inline tpu_inference/layers/common/attention_interface.py:1]: Are the changes in this fils accidental?

**Review by BirdsOfAFthr — APPROVED** [review]

File: tpu_inference/layers/common/attention_interface.py
Comment (karan) [reply tpu_inference/layers/common/attention_interface.py:1]: Yes - I had reverted them this morning

**Review by gxd3 — APPROVED** [review]
